### ISSUE TITLE
[otbn] Mismatch between RTL and ISS

### ISSUE THREAD

**Issue by vrozic** [lowRISC/opentitan#17154]
### Description

Verilator simulation sometimes reports a mismatch between ISS and RTL.
To reproduce the error:
- `bazel build //sw/...`
- `fusesoc --cores-root=. run --target=sim --setup --build lowrisc:ip:otbn_top_sim`
- `./build/lowrisc_ip_otbn_top_sim_0.1/sim-verilator/Votbn_top_sim --trace  --load-elf /path/to/code-snippets/mul384.elf`

Reported error is: `ERROR: At time 420, insn_cnt != otbn_model_insn_cnt (0xd != 0xc).`

**Comment by ctopal** [comment]
I believe this is happening because we are not initialising the wider registers same way as the RTL in the model. However, there's still a chance that this is related with the tracer as well. I'm thinking like this because wr_en is not `0x11` but `0x01` and logs are matching for the first half...

I believe we don't want to deal with setting "trashed/uninitialised" memory same way as RTL in ISS, so I'm thinking about coming up with a fix in the `otbn_tracer.sv` so that we'd only report enabled part in RTL side.

**Comment by ctopal** [comment]
TLDR: Data written to registers reports differently in RTL vs ISS as their initial values mismatch in the case of writing to only a part of said register.

I believe this is going to be tricky to solve and it's related with how we do tracing. 

The actual error happens in: `bn.mulqacc.so w4.L, w0.1, w2.0, 64` instruction of `mul384.s` test but essentially can be seen in bunch of other tests in which we write to a part of the uninitialised registers.

Error Message:
```
ERROR: Mismatch between RTL and ISS trace entries: Final values of ISS and RTL don't match for key `w04'.
  RTL entry is:
    > w04: 0xda3ffcfa_99603a6e_20f36dea_00178ebb_c4df807f_413b6763_921c8782_f7f42166
  ISS entry is:
    > w04: 0x00000000_00000000_00000000_00000000_c4df807f_413b6763_921c8782_f7f42166
```

Current way of RTL tracing work is that we set the `wr_data` using the old value in the register and the new one that is calculated and use write-enable signal to decide which one to use. ISS tries to do the same thing but as it's initial value is 0 and not a random value it's not matching what RTL does. In RTL we do have a random value in the `w4` initially because of the out-of-reset secure wipe. 

How RTL handles it: this:https://github.com/lowRISC/opentitan/blob/0cba2acca51c0132ce621b35dac1728ae8803c07/hw/ip/otbn/dv/tracer/rtl/otbn_trace_if.sv#L170-L195

How ISS handles it:
https://github.com/lowRISC/opentitan/blob/0cba2acca51c0132ce621b35dac1728ae8803c07/hw/ip/otbn/dv/otbnsim/sim/insn.py#L730-L738

Note that I think this would work without a problem if we were to initialise `w4`.

**Comment by ctopal** [comment]
@GregAC + @andreaskurth for visibility

**Comment by andreaskurth** [comment]
Thanks @ctopal, I think your suggestion to initialize registers in SW to prevent this RTL/ISS mismatch makes sense. The RTL behaves correctly, and while the ISS could be tuned to behave more closely like RTL (e.g., using an undefined value for registers after secure wipe instead of all zero), I think that's too much effort for little gain.

So I suggest to
- Check the OTBN doc to ensure it contains a statement similar to "After secure wipe, the value of all registers is undefined. Make sure to initialize any register in software before using its value."
- Amend tests that use uninitialized registers, including use cases where software does not observe the uninitialized (or partially initialized) register value but the ISS does observe it.
